data/transactions/logic/eval.go

```diff
@@ -3911,10 +4262,96 @@ func (cx *EvalContext) stackIntoTxnField(sv stackValue, fs txnFieldSpec, txn *tr
 	case FreezeAssetFrozen:
 		txn.AssetFrozen, err = sv.bool()
 
-	// appl needs to wait. Can't call AVM from AVM.
-
+	// ApplicationCall
+	case ApplicationID:
+		txn.ApplicationID, err = cx.availableApp(sv)
```

Comment:
this might return self and we do not allow calling self, not really matter but might fail earlier here

Reply:
In order to make the spec clear (we are trying to document whether things fail when fields are set, or when submit is called). Rather than have a lot of special cases, we say that we only fail when a field is set if it's the wrong type, or if the "thing" (account, app, asa) isn't "available" (in the arrays or accessible by a new rule).

Everything else fails at itxn_submit time.